File: android/app/src/main/java/com/voiceguard/android/analysis/Heuristic.kt
package com.voiceguard.android.analysis

import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min

object HeuristicDetector {
    private fun clamp(x: Float, lo: Float, hi: Float): Float {
        return min(max(x, lo), hi)
    }

    private fun sigmoid(x: Float): Float {
        return (1.0f / (1.0f + exp(-x)))
    }

    fun pFake(indicators: Indicators): Pair<Float, List<String>> {
        val hfRatio = indicators.hfEnergyRatio
        val rolloffHz = indicators.spectralRolloffHz
        val flatness = indicators.spectralFlatness
        val bandwidthHz = indicators.spectralBandwidthHz
        val centroidHz = indicators.spectralCentroidHz
        val zcr = indicators.zcr

        val cutoffScore = clamp((0.06f - hfRatio) / 0.06f, 0.0f, 1.0f)
        val rolloffScore = clamp((4200.0f - rolloffHz) / 4200.0f, 0.0f, 1.0f)
        val flatScore = clamp((0.12f - flatness) / 0.12f, 0.0f, 1.0f)
        val bandwidthScore = clamp((1800.0f - bandwidthHz) / 1800.0f, 0.0f, 1.0f)
        val centroidScore = clamp((1700.0f - centroidHz) / 1700.0f, 0.0f, 1.0f)
        val zcrScore = clamp((0.04f - zcr) / 0.04f, 0.0f, 1.0f)

        val raw = 0.36f * cutoffScore +
            0.27f * rolloffScore +
            0.13f * flatScore +
            0.12f * bandwidthScore +
            0.07f * centroidScore +
            0.05f * zcrScore
        val pFake = sigmoid((raw - 0.34f) * 7.0f)

        return clamp(pFake, 0.0f, 1.0f) to reasons(indicators)
    }

    fun reasons(indicators: Indicators): List<String> {
        val hfRatio = indicators.hfEnergyRatio
        val rolloffHz = indicators.spectralRolloffHz
        val flatness = indicators.spectralFlatness
        val bandwidthHz = indicators.spectralBandwidthHz
        val centroidHz = indicators.spectralCentroidHz
        val zcr = indicators.zcr

        val reasons = mutableListOf<String>()
        if (hfRatio < 0.02f) {
            reasons.add("Низкая доля ВЧ энергии (возможен срез/кодек/TTS)")
        }
        if (rolloffHz < 3200.0f) {
            reasons.add("Низкий spectral roll-off (возможен срез ВЧ)")
        }
        if (flatness < 0.08f) {
            reasons.add("Низкая spectral flatness (слишком стерильный спектр)")
        }
        if (bandwidthHz < 1800.0f) {
            reasons.add("Слишком узкий спектр (ограничение полосы)")
        }
        if (centroidHz < 1700.0f) {
            reasons.add("Смещение энергии в НЧ (мутный/перефильтрованный звук)")
        }
        if (zcr < 0.03f) {
            reasons.add("Низкий ZCR (слишком ровный/тональный сигнал)")
        }
        return reasons
    }
}

File: voiceguard/inference/heuristic.py
from __future__ import annotations

import math
from typing import Mapping, Tuple


def _clamp(x: float, lo: float, hi: float) -> float:
    return float(min(max(x, lo), hi))


def _sigmoid(x: float) -> float:
    return float(1.0 / (1.0 + math.exp(-x)))


def heuristic_p_fake(indicators: Mapping[str, float]) -> Tuple[float, list[str]]:
    hf_ratio = float(indicators.get("hf_energy_ratio", 0.0))
    rolloff_hz = float(indicators.get("spectral_rolloff_hz", 0.0))
    flatness = float(indicators.get("spectral_flatness", 0.0))

    # Very rough, demo-only heuristic:
    # - many synthetic/over-compressed samples show reduced HF energy / rolloff
    # - overly "sterile" spectrum can have low flatness
    cutoff_score = _clamp((0.06 - hf_ratio) / 0.06, 0.0, 1.0)
    rolloff_score = _clamp((4200.0 - rolloff_hz) / 4200.0, 0.0, 1.0)
    flat_score = _clamp((0.12 - flatness) / 0.12, 0.0, 1.0)

    raw = 0.50 * cutoff_score + 0.35 * rolloff_score + 0.15 * flat_score
    p_fake = _sigmoid((raw - 0.35) * 7.0)

    reasons = heuristic_reasons(indicators)

    return float(_clamp(p_fake, 0.0, 1.0)), reasons


def heuristic_reasons(indicators: Mapping[str, float]) -> list[str]:
    hf_ratio = float(indicators.get("hf_energy_ratio", 0.0))
    rolloff_hz = float(indicators.get("spectral_rolloff_hz", 0.0))
    flatness = float(indicators.get("spectral_flatness", 0.0))

    reasons: list[str] = []
    if hf_ratio < 0.02:
        reasons.append("низкая доля энергии в ВЧ (возможен срез/кодек/TTS)")
    if rolloff_hz < 3200.0:
        reasons.append("низкий spectral roll-off (возможен срез ВЧ)")
    if flatness < 0.08:
        reasons.append("низкая spectral flatness (слишком 'стерильно/тонально')")
    return reasons
